src/Sulu/Component/Content/Document/SynchronizationManager.php

```diff
@@ -0,0 +1,252 @@
+use Sulu\Component\DocumentManager\DocumentManagerRegistryInterface;
+
+/**
+ * The synchronization manager handles the synchronization of documents
+ * from the DEFAULT document manger to the PUBLISH document manager.
+ *
+ * NOTE: In the future multiple document managers may be supported.
+ *
+ * NOTE: Much of the logic in this class is about ensuring that corresponding
+ *       nodes exist in the PDM - this could be simplified (and made much more
+ *       efficient) if we were to use Jackalope observation to implement Event
+ *       listeners at the PHPCR level, however this feature is currently not
+ *       supported.
+ *
+ *       See: https://github.com/jackalope/jackalope/pull/241
+ *
+ * TODO: Explcitly inject publish and default document managers?
+ */
+class SynchronizationManager
+{
+    /**
+     * @var DocumentManagerRegistryInterface
+     */
+    private $registry;
+
+    /**
+     * @var PropertyEncoder
+     */
+    private $encoder;
+
+    /**
+     * @var string
+     */
+    private $publishManagerName;
+
+    /**
+     * @var DocumentRegistrator
+     */
+    private $registrator;
+
+    public function __construct(
+        DocumentManagerRegistry $registry,
+        PropertyEncoder $encoder,
+        $publishManagerName,
+        $registrator = null
+    ) {
+        $this->registry = $registry;
+        $this->publishManagerName = $publishManagerName;
+        $this->encoder = $encoder;
+        $this->registrator = $registrator ?: new DocumentRegistrator(
+            $registry->getManager(),
+            $registry->getManager($this->publishManagerName)
+        );
+    }
+
+    /**
+     * Return the publish document manager (PDM).
+     *
+     * This should be the only class that is aware of the PDM name. By having
+     * this method we can be sure that whatever the PDM is, the PDM is always
+     * the PDM.
+     *
+     * NOTE: This is used only by the synchronization subscriber in order
+     *       to "flush" the PDM.
+     *
+     * @return DocumentManagerInterface
+     */
+    public function getPublishDocumentManager()
+    {
+        return $this->registry->getManager($this->publishManagerName);
+    }
+
+    /**
+     * Synchronize a document and any other documents which are
+     * associated with it and should also be published.
+     *
+     * All document managers involved will be FLUSHED after
+     * the operation has completed.
+     *
+     * @param SynchronizeBehavior $document
+     */
+    public function synchronizeFull(SynchronizeBehavior $document, array $options = [])
+    {
+        // get the default managerj
+        $defaultManager = $this->registry->getManager();
+        $publishManager = $this->getPublishDocumentManager();
+
+        // if the  emitting manager is the publish manager, stop we
+        // don't want to sync from the publish workspace!
+        //
+        // TODO: this should be removed by implementing per-document-manager
```

Comment:
Remove this TODO. We check if the publish and default managers are the same because if they are then there is nothing to sync to.